File: 563/563.py
# @brief:  leetcode-563. 二叉树的坡度
# @envir:  python 2.7
# @date:   2019/5/11
# @author: luhao

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.ans = 0
        
    def sumNode(self, root):
        # 计算子树的节点之和
        if not root:
            return 0
        left = self.sumNode(root.left)
        right = self.sumNode(root.right)
        self.ans += abs(left-right)
        return root.val + left + right
    
    
    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.sumNode(root)
        return self.ans

File: 563/563.cpp
/*!***********************
 * @brief:  leetcode-563. 二叉树的坡度
 * @date:   2019/5/11
 * @envir:  gcc 7.4.0
 * @author: luhao
 * ***********************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int ans=0;
    int sumNode(TreeNode* root) {
        if (!root)
            return 0;
        int left = sumNode(root->left);
        int right = sumNode(root->right);
        ans += (right>left)?right-left:left-right;
        return root->val + left + right;
    }
    
    int findTilt(TreeNode* root) {
        if (!root)
            return 0;
        sumNode(root);
        return ans;
    }
};
